### PR TITLE
concurrent form processing

### PR DESCRIPTION
**Pull request by jmtroth0**
Motivation:
It seems main form processing is currently taking a decent chunk of [time](https://app.datadoghq.com/dash/773587/jordans-checks?live=true&page=0&is_auto=false&from_ts=1523138804551&to_ts=1525730804551&tile_size=xs). Many seem to take < 0.1 seconds, but those seem to be the small domains. Big ones with complex apps take much longer, for instance a load testing domain `hsph-load-testing`, for which migrating 580 forms took 30 minutes.

It's seeming like 0.5s/form is a reasonable estimate. This is fine while all the domains I am doing are under ~50000, which would take 7 hours assuming 0.5s (`50000 * .5 / 60 / 60`), but if we are still thinking of allowing for down time, `rec`, for instance, isn't going to be happy. Correct me if I'm wrong, but I looked at their submission report and it seems like they have 1.5 million over the past 3 months, and they've been around for awhile. This is 30x the above numbers over a short period of time, which is untenable.

Solution?
This is one possible aid that I timeboxed a prototype to this afternoon. Since it seems like processing forms into a case is what restricts us to sequential processing, this keeps it sequential by case.

I still want to write tests/hear about edge cases, but I wanted an early review to make sure my assumptions on this were also correct.

@snopoke @proteusvacuum @millerdev

### CODE REVIEW

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 1]
```diff
@@ -110,11 +117,41 @@ def _process_main_forms(self):
             form_received = wrapped_form.received_on
             assert last_received_on <= form_received
             last_received_on = form_received
-            try:
-                self._migrate_form_and_associated_models(wrapped_form)
-            except Exception:
-                self.log_error("Unable to migrate form: {}".format(change.id))
-                raise
+
+            case_doc = wrapped_form.form.get('case')
```
Comment (snopoke) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: forms can update multiple cases. You can use `get_case_ids_from_form` to get all cases touched by a form.

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 2]
```diff
@@ -617,3 +654,71 @@ def commit_migration(domain_name):
         assert should_use_sql_backend(domain_name)
     datadog_counter("commcare.couch_sql_migration.total_committed")
     _logger.info("committed migration for {}".format(domain))
+
+
+class PartiallyLockingQueue(object):
+    def __init__(self, key_base):
+        self.queue_by_id = defaultdict(deque)
+        self.key_base = "%s.%%s.%%s" % key_base  # <key_base>.<lock_type>.<obj_id>
+        self.redis = get_redis_client()
```
Comment (snopoke) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: since this is only running on a single machine can we not just use local locking?

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 3]
```diff
@@ -617,3 +654,71 @@ def commit_migration(domain_name):
...
+        if not value:
+            return False
+        if value > time():
+            return True
+        self.release_lock(obj_id)
+        raise LockTimeoutException("Object {} should not be taking so long for this action".format(obj_id))
+
+    def set_lock(self, obj_id):
+        expiry_time = self.get_expiry_time()
+        value = self.redis.get_or_set(self.get_obj_lock_key(obj_id), expiry_time)
```
Comment (snopoke) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: why not just use actual redis locks?

**Review by snopoke — COMMENTED** [review]
Nice idea. One complexity is that forms can update multiple cases. In the form processing code we handle this by blocking submissions if they touch a case that's being processed.

**Comment by jmtroth0** [comment]
@snopoke updated some things:

1) Obj locking can be managed locally rather than in redis, so I removed it 6c8c3a8
2) I made it so the queue can handle multiple case ids for each form 7095fc1

The rest of the new stuff (4eac095 and onward) is cleanup.

I had a couple of ideas for 2, but this seemed the clearest as you can see from this drawing. This was option 2, below the solid black line
img_9487

Basically, I now am tracking 2 dicts. (1) is the same as was there before, tracking from `case_id` to a `queue<form>`, and (2) now also allows tracking from `form_id` to a `list<case_ids>` that are associated with it. This made removal easier as well as simplifying the `try_obj` function. I am also adding some docstrings and comments to explain what each of the functions does

**Comment by jmtroth0** [comment]
Added some inline documentation, will also add tests tmw morning

**Comment by snopoke** [comment]
Looks good. Build failing though (see below). 
```
ERROR: corehq.apps.couch_sql_migration.tests.test_migration:MigrationTestCase.test_basic_case_migration_case_name
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/tests/test_migration.py", line 401, in test_basic_case_migration_case_name
    self._do_migration_and_assert_flags(self.domain_name)
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/tests/test_migration.py", line 69, in _do_migration_and_assert_flags
    self._do_migration(domain)
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/tests/test_migration.py", line 66, in _do_migration
    call_command('migrate_domain_from_couch_to_sql', domain, MIGRATE=True, no_input=True)
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "/vendor/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py", line 71, in handle
    do_couch_to_sql_migration(domain, with_progress=not self.no_input, debug=self.debug)
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/couchsqlmigration.py", line 63, in do_couch_to_sql_migration
    CouchSqlDomainMigrator(domain, with_progress=with_progress, debug=debug).migrate()
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/couchsqlmigration.py", line 96, in migrate
    self._process_main_forms()
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/couchsqlmigration.py", line 126, in _process_main_forms
    if self.queues.try_obj(case_ids, wrapped_form):
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/couchsqlmigration.py", line 709, in try_obj
    self.add_item(lock_ids, queue_obj)
  File "/mnt/commcare-hq-ro/corehq/apps/couch_sql_migration/couchsqlmigration.py", line 767, in add_item
    self.lock_ids_by_queue_id[self.get_queue_obj_id(queue_obj)] = lock_ids
TypeError: get_queue_obj_id() takes exactly 2 arguments (1 given)
```

**Review by snopoke — APPROVED** [review]

**Comment by snopoke** [comment]
I think you'll need to add this command to the list here: https://github.com/dimagi/commcare-hq/blob/7f0593cb35e4251799ee4b4ad44d3e8e79a97cc9/manage.py#L112

Also build still failing.

File: corehq/apps/couch_sql_migration/tests/test_migration.py [hunk 4]
```diff
@@ -623,3 +623,111 @@ def _validate_ledger_data(self, state_dict, expected):
...
+        self.queues = PartiallyLockingQueue()
+
+    def _add_to_queues(self, queue_obj_id, lock_ids):
+        self.queues.add_item(lock_ids, DummyObject(queue_obj_id))
+        self._check_queue_dicts(queue_obj_id, lock_ids, -1)
+
+    def _check_queue_dicts(self, queue_obj_id, lock_ids, location=None, present=True):
+        """
+        if location is None, it looks anywhere. If it is an int, it'll look in that spot
+        present determines whether it's expected to be in the queue_by_lock_id or not
```
Comment (millerdev) [inline corehq/apps/couch_sql_migration/tests/test_migration.py]: `present` is ignored if `bool(location) == True`. Does that matter?

**Review by millerdev — COMMENTED** [review]
Bunch of nitpicky comments. Not sure if any of them are important enough to require changes though.

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 5]
```diff
@@ -631,3 +669,154 @@ def commit_migration(domain_name):
...
+        queue it would sit in
+
+        Returns :boolean: True if it acquired the lock, False if it was added to queue
+        """
+        if self.check_lock(lock_ids):  # if it's currently locked, it can't acquire the lock
+            self.add_item(lock_ids, queue_obj)
+            return False
+        for lock_id in lock_ids:  # if other objs are waiting for the same locks, it has to wait
+            queue = self.queue_by_lock_id[lock_id]
+            if queue:
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: because `self.queue_by_lock_id` is a `defaultdict`, it'll always be an empty `deque` rather than `None`

File: corehq/apps/couch_sql_migration/tests/test_migration.py [hunk 6]
```diff
@@ -623,3 +623,111 @@ def _validate_ledger_data(self, state_dict, expected):
...
+    def setUp(self):
+        self.queues = PartiallyLockingQueue()
+
+    def _add_to_queues(self, queue_obj_id, lock_ids):
+        self.queues.add_item(lock_ids, DummyObject(queue_obj_id))
+        self._check_queue_dicts(queue_obj_id, lock_ids, -1)
+
+    def _check_queue_dicts(self, queue_obj_id, lock_ids, location=None, present=True):
+        """
+        if location is None, it looks anywhere. If it is an int, it'll look in that spot
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/tests/test_migration.py]: good catch

File: corehq/apps/couch_sql_migration/tests/test_migration.py (hunk 4 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/tests/test_migration.py]: Nope, not for my use case. Since it's easy enough, I might as well make it matter though

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 7]
```diff
@@ -631,3 +669,154 @@ def commit_migration(domain_name):
...
+            first_in_all_queues = True
+            for lock_id in lock_ids:
+                first_in_queue = self.queue_by_lock_id[lock_id][0]  # can assume there always will be one
+                if not self.get_queue_obj_id(first_in_queue) == peeked_id:
+                    first_in_all_queues = False
+                    break
+            if not first_in_all_queues:
+                continue
+
+            if self.set_lock(lock_ids):
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: I sort of like the separation actually
